Data too long for column 'DETAILS_JSON'

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

oidc

### Describe the bug

When spamming keycloak-javascript-adapter initiated logins by refreshing my frontend many times in a row, I can find the following exception in the Keycloak logs. The user is being redirected to the "500 internal server error" Keycloak website. 

Logs: 

```
2023-02-22 11:27:16,457 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-521) SQL Error: 1406, SQLState: 22001
2023-02-22 11:27:16,457 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-521) Data truncation: Data too long for column 'DETAILS_JSON' at row 1
2023-02-22 11:27:16,458 INFO  [org.hibernate.engine.jdbc.batch.internal.AbstractBatchImpl] (executor-thread-521) HHH000010: On release of batch it still contained JDBC statements
2023-02-22 11:27:16,465 WARN  [com.arjuna.ats.arjuna] (executor-thread-521) ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff6443074d:a937:63ea0e5c:1beeac, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@2f772455 >: javax.persistence.PersistenceException: org.hibernate.exception.DataException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1411)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:489)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3303)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2438)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:449)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:356)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:76)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:360)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:91)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1295)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:128)
	at io.quarkus.narayana.jta.runtime.CDIDelegatingTransactionManager.commit(CDIDelegatingTransactionManager.java:105)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:90)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler.close(TransactionalSessionHandler.java:64)
	at org.keycloak.quarkus.runtime.integration.jaxrs.TransactionalResponseFilter.filter(TransactionalResponseFilter.java:47)
	at org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl.filter(ContainerResponseContextImpl.java:367)
	at org.jboss.resteasy.core.ServerResponseWriter.executeFilters(ServerResponseWriter.java:252)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:101)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:74)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeResponse(SynchronousDispatcher.java:594)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:524)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.hibernate.exception.DataException: could not execute statement
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:52)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3937)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1407)
	... 54 more
Caused by: com.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: Data too long for column 'DETAILS_JSON' at row 1
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:104)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:916)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1061)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1009)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1320)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:994)
	at com.mysql.cj.jdbc.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:596)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 66 more
```

### Version

20.0.5

### Expected behavior

The Keycloak server does not get internal server errors when the user attempts to login multiple times in a row

### Actual behavior

Refreshing/requesting the token of the user does end up in a 500 internal server error response

### How to Reproduce?

Use a frontend library to initiate the OIDC login automatically when you open the frontend website.
Refresh the website in your browser in a quick succession after the login was successful using the same session (do not use private windows or clear cookies in-between). 
After some attempts, the abovementioned exception will be thrown. 

### Anything else?

_No response_

--- a/docs/documentation/server_admin/topics/events/login.adoc
+++ b/docs/documentation/server_admin/topics/events/login.adoc
@@ -171,17 +171,3 @@ You can exclude events by using the `--spi-events-listener-email-exclude-events`
 kc.[sh|bat] --spi-events-listener-email-exclude-events=UPDATE_TOTP,REMOVE_TOTP
 ----
 
-You can set a maximum length of each Event detail in the database by using the `--spi-events-store-jpa-max-detail-length` argument. This setting is useful if a detail (for example, redirect_uri) is long. For example:
-
-[source,bash]
-----
-kc.[sh|bat] --spi-events-store-jpa-max-detail-length=1000
-----
-
-Also you can set a maximum length of all Event's details by using the `--spi-events-store-jpa-max-field-length` argument. This setting is useful if you want to adhere to the underlying storage limitation. For example:
-
-[source,bash]
-----
-kc.[sh|bat] --spi-events-store-jpa-max-field-length=2500
-----
-

--- a/docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc
+++ b/docs/documentation/upgrading/topics/keycloak/changes-23_0_0.adoc
@@ -141,3 +141,7 @@ Relying on RESTEasy Classic is not longer an option because it is not available 
 = Partial export requires manage-realm permission
 
 The endpoint `POST {keycloak server}/realms/{realm}/partial-export` and the corresponding action in the admin console now require `manage-realm` permission for execution instead of `view-realm`. This endpoint exports the realm configuration into a JSON file and the new permission is more appropriate. The parameters `exportGroupsAndRoles` and `exportClients`, which include the realm groups/roles and clients in the export respectively, continue managing the same permissions (`query-groups` and `view-clients`).
+
+= Removal of the options to trim the event's details length
+
+Since this release, Keycloak supports long value for `EventEntity` details column. Therefore, it no longer supports options for trimming event detail length `--spi-events-store-jpa-max-detail-length` and `--spi-events-store-jpa-max-field-length`.

--- a/model/jpa/src/main/java/org/keycloak/events/jpa/AdminEventEntity.java
+++ b/model/jpa/src/main/java/org/keycloak/events/jpa/AdminEventEntity.java
@@ -60,7 +60,7 @@ public class AdminEventEntity {
     @Column(name="RESOURCE_PATH")
     private String resourcePath;
 
-    @Column(name="REPRESENTATION", length = 25500)
+    @Column(name="REPRESENTATION")
     private String representation;
 
     @Column(name="ERROR")

--- a/model/jpa/src/main/java/org/keycloak/events/jpa/EventEntity.java
+++ b/model/jpa/src/main/java/org/keycloak/events/jpa/EventEntity.java
@@ -57,8 +57,12 @@ public class EventEntity {
     @Column(name="ERROR")
     private String error;
 
+    // This is the legacy field which is kept here to be able to read old events without the need to migrate them
     @Column(name="DETAILS_JSON", length = 2550)
     private String detailsJson;
+
+    @Column(name="DETAILS_JSON_LONG_VALUE")
+    private String detailsJsonLongValue;
 
     public String getId() {
         return id;
@@ -133,11 +137,11 @@ public class EventEntity {
     }
 
     public String getDetailsJson() {
-        return detailsJson;
+        return detailsJsonLongValue != null ? detailsJsonLongValue : detailsJson;
     }
 
     public void setDetailsJson(String detailsJson) {
-        this.detailsJson = detailsJson;
+        this.detailsJsonLongValue = detailsJson;
     }
 
 }

--- a/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProvider.java
+++ b/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProvider.java
@@ -39,7 +39,6 @@ import org.keycloak.models.utils.KeycloakModelUtils;
 import jakarta.persistence.EntityManager;
 import jakarta.persistence.TypedQuery;
 import java.io.IOException;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
@@ -57,14 +56,10 @@ public class JpaEventStoreProvider implements EventStoreProvider {
 
     private final KeycloakSession session;
     private final EntityManager em;
-    private final int maxDetailLength;
-    private final int maxFieldLength;
 
-    public JpaEventStoreProvider(KeycloakSession session, EntityManager em, int maxDetailLength, int maxFieldLength) {
+    public JpaEventStoreProvider(KeycloakSession session, EntityManager em) {
         this.session = session;
         this.em = em;
-        this.maxDetailLength = maxDetailLength;
-        this.maxFieldLength = maxFieldLength;
     }
 
     @Override
@@ -165,38 +160,11 @@ public class JpaEventStoreProvider implements EventStoreProvider {
         eventEntity.setIpAddress(event.getIpAddress());
         eventEntity.setError(event.getError());
         try {
-            if (maxDetailLength > 0 && event.getDetails() != null) {
-                Map<String, String> result = new HashMap<>(event.getDetails());
-                result.entrySet().forEach(t -> t.setValue(trimToMaxDetailLength(t.getValue())));
-
-                eventEntity.setDetailsJson(trimToMaxFieldLength(mapper.writeValueAsString(result)));
-            } else {
-                eventEntity.setDetailsJson(mapper.writeValueAsString(event.getDetails()));
-            }
+            eventEntity.setDetailsJson(mapper.writeValueAsString(event.getDetails()));
         } catch (IOException ex) {
             logger.error("Failed to write log details", ex);
         }
         return eventEntity;
-    }
-
-    private String trimToMaxDetailLength(String detail) {
-        if (detail != null && detail.length() > maxDetailLength) {
-            logger.warnf("Detail '%s' will be truncated.", detail);
-            // (maxDetailLength - 3) takes "..." into account
-            return detail.substring(0, maxDetailLength - 3).concat("...");
-        } else {
-            return detail;
-        }
-    }
-
-    private String trimToMaxFieldLength(String field) {
-        if (maxFieldLength > 0 && field != null && field.length() > maxFieldLength) {
-            logger.warnf("Field '%s' will be truncated.", field);
-            // (maxFieldLength - 3) takes "..." into account
-            return field.substring(0, maxFieldLength - 3).concat("...");
-        } else {
-            return field;
-        }
     }
 
     static Event convertEvent(EventEntity eventEntity) {
@@ -234,8 +202,8 @@ public class JpaEventStoreProvider implements EventStoreProvider {
         adminEventEntity.setResourcePath(adminEvent.getResourcePath());
         adminEventEntity.setError(adminEvent.getError());
         
-        if(includeRepresentation) {
-            adminEventEntity.setRepresentation(trimToMaxFieldLength(adminEvent.getRepresentation()));
+        if (includeRepresentation) {
+            adminEventEntity.setRepresentation(adminEvent.getRepresentation());
         }
         return adminEventEntity;
     }
@@ -255,7 +223,7 @@ public class JpaEventStoreProvider implements EventStoreProvider {
         adminEvent.setResourcePath(adminEventEntity.getResourcePath());
         adminEvent.setError(adminEventEntity.getError());
         
-        if(adminEventEntity.getRepresentation() != null) {
+        if (adminEventEntity.getRepresentation() != null) {
             adminEvent.setRepresentation(adminEventEntity.getRepresentation());
         }
         return adminEvent;

--- a/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProviderFactory.java
+++ b/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProviderFactory.java
@@ -32,21 +32,15 @@ import org.keycloak.storage.datastore.PeriodicEventInvalidation;
 public class JpaEventStoreProviderFactory implements EventStoreProviderFactory, InvalidationHandler {
 
     public static final String ID = "jpa";
-    private int maxDetailLength;
-    private int maxFieldLength;
 
     @Override
     public EventStoreProvider create(KeycloakSession session) {
         JpaConnectionProvider connection = session.getProvider(JpaConnectionProvider.class);
-        return new JpaEventStoreProvider(session, connection.getEntityManager(), maxDetailLength, maxFieldLength);
+        return new JpaEventStoreProvider(session, connection.getEntityManager());
     }
 
     @Override
     public void init(Config.Scope config) {
-        maxDetailLength = config.getInt("max-detail-length", -1);
-        maxFieldLength = config.getInt("max-field-length", -1);
-        if (maxDetailLength != -1 && maxDetailLength < 3) throw new IllegalArgumentException("max-detail-length cannot be less that 3.");
-        if (maxFieldLength != -1 && maxFieldLength < 3) throw new IllegalArgumentException("max-field-length cannot be less that 3.");
     }
 
     @Override

--- a/model/jpa/src/main/resources/META-INF/jpa-changelog-23.0.0.xml
+++ b/model/jpa/src/main/resources/META-INF/jpa-changelog-23.0.0.xml
@@ -30,4 +30,11 @@
         <dropColumn tableName="COMPONENT_CONFIG" columnName="VALUE"/>
         <renameColumn tableName="COMPONENT_CONFIG" oldColumnName="VALUE_NEW" newColumnName="VALUE" columnDataType="NCLOB"/>
     </changeSet>
+
+    <changeSet author="keycloak" id="23.0.0-17258">
+        <addColumn tableName="EVENT_ENTITY">
+            <column name="DETAILS_JSON_LONG_VALUE" type="NCLOB" />
+        </addColumn>
+    </changeSet>
+
 </databaseChangeLog>

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/event/AdminEventTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/event/AdminEventTest.java
@@ -17,6 +17,7 @@
 
 package org.keycloak.testsuite.admin.event;
 
+import org.apache.commons.lang3.RandomStringUtils;
 import org.junit.Before;
 import org.junit.Test;
 import org.keycloak.admin.client.resource.RealmResource;
@@ -45,6 +46,7 @@ import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
 import static org.junit.Assert.assertNull;
 import static org.keycloak.testsuite.auth.page.AuthRealm.MASTER;
 
@@ -165,6 +167,23 @@ public class AdminEventTest extends AbstractEventTest {
         assertThat(realm.getAdminEvents(null, null, null, null, null, null, null, null, null, null).size(), is(equalTo(100)));
         assertThat(realm.getAdminEvents(null, null, null, null, null, null, null, null, 0, 105).size(), is(equalTo(105)));
         assertThat(realm.getAdminEvents(null, null, null, null, null, null, null, null, 0, 1000).size(), is(greaterThanOrEqualTo(110)));
+    }
+
+    @Test
+    public void adminEventRepresentationLenght() {
+        RealmResource realm = adminClient.realms().realm("test");
+        AdminEventRepresentation event = new AdminEventRepresentation();
+        event.setOperationType(OperationType.CREATE.toString());
+        event.setAuthDetails(new AuthDetailsRepresentation());
+        event.setRealmId(realm.toRepresentation().getId());
+        String longValue = RandomStringUtils.random(30000, true, true);
+        event.setRepresentation(longValue);
+
+        testingClient.testing("test").onAdminEvent(event, true);
+        List<AdminEventRepresentation> adminEvents = realm.getAdminEvents(null, null, null, null, null, null, null, null, null, null);
+
+        assertThat(adminEvents, hasSize(1));
+        assertThat(adminEvents.get(0).getRepresentation(), equalTo(longValue));
     }
 
     @Test

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/AdminEventQueryTest.java
@@ -17,8 +17,11 @@
 package org.keycloak.testsuite.model.events;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
 
+import org.apache.commons.lang3.RandomStringUtils;
 import org.junit.Test;
 import org.keycloak.common.ClientConnection;
 import org.keycloak.events.EventStoreProvider;
@@ -38,6 +41,7 @@ import org.keycloak.testsuite.model.KeycloakModelTest;
 import org.keycloak.testsuite.model.RequireProvider;
 
 import java.util.List;
+import java.util.Map;
 import java.util.stream.Collectors;
 
 @RequireProvider(EventStoreProvider.class)
@@ -109,6 +113,30 @@ public class AdminEventQueryTest extends KeycloakModelTest {
             assertThat(adminEventsDesc.size(), is(2));
             assertThat(adminEventsDesc.get(0).getOperationType(), is(OperationType.DELETE));
             assertThat(adminEventsDesc.get(1).getOperationType(), is(OperationType.CREATE));
+            return null;
+        });
+    }
+
+        @Test
+    public void testAdminEventRepresentationLongValue() {
+        String longValue = RandomStringUtils.random(30000, true, true);
+
+        withRealm(realmId, (session, realm) -> {
+            
+            AdminEvent event = createClientEvent(realm, OperationType.CREATE);
+            event.setRepresentation(longValue);
+
+            session.getProvider(EventStoreProvider.class).onEvent(event, true);
+
+            return null;
+        });
+
+        withRealm(realmId, (session, realm) -> {
+            List<AdminEvent> events = session.getProvider(EventStoreProvider.class).createAdminQuery().realm(realmId).getResultStream().collect(Collectors.toList());
+            assertThat(events, hasSize(1));
+
+            assertThat(events.get(0).getRepresentation(), equalTo(longValue));
+
             return null;
         });
     }

--- a/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/EventQueryTest.java
+++ b/testsuite/model/src/test/java/org/keycloak/testsuite/model/events/EventQueryTest.java
@@ -27,9 +27,13 @@ import org.keycloak.models.RealmModel;
 import org.keycloak.testsuite.model.KeycloakModelTest;
 import org.keycloak.testsuite.model.RequireProvider;
 import java.util.List;
+import java.util.Map;
 import java.util.stream.Collectors;
+import org.apache.commons.lang3.RandomStringUtils;
 import org.junit.Test;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
 
 /**
@@ -130,6 +134,38 @@ public class EventQueryTest extends KeycloakModelTest {
             assertThat(eventsDesc.get(0).getUserId(), is("u2"));
             assertThat(eventsDesc.get(1).getUserId(), is("u1"));
 
+            return null;
+        });
+    }
+
+    @Test
+    public void testEventDetailsLongValue() {
+        String v1 = RandomStringUtils.random(1000, true, true);
+        String v2 = RandomStringUtils.random(1000, true, true);
+        String v3 = RandomStringUtils.random(1000, true, true);
+        String v4 = RandomStringUtils.random(1000, true, true);
+
+        withRealm(realmId, (session, realm) -> {
+
+            Map<String, String> details = Map.of("k1", v1, "k2", v2, "k3", v3, "k4", v4);
+
+            Event event = createAuthEventForUser(session, realm, "u1");
+            event.setDetails(details);
+
+            session.getProvider(EventStoreProvider.class).onEvent(event);
+
+            return null;
+        });
+
+        withRealm(realmId, (session, realm) -> {
+            List<Event> events = session.getProvider(EventStoreProvider.class).createQuery().realm(realmId).getResultStream().collect(Collectors.toList());
+            assertThat(events, hasSize(1));
+            Map<String, String> details = events.get(0).getDetails();
+
+            assertThat(details.get("k1"), equalTo(v1));
+            assertThat(details.get("k2"), equalTo(v2));
+            assertThat(details.get("k3"), equalTo(v3));
+            assertThat(details.get("k4"), equalTo(v4));
             return null;
         });
     }

